sympy/core/numbers.py

```diff
@@ -4436,8 +4436,107 @@ def as_base_exp(self):
     def _mpc_(self):
         return (Float(0)._mpf_, Float(1)._mpf_)
 
+
 I = S.ImaginaryUnit
 
+
+def equal_valued(x, y):
+    """Compare expressions treating plain floats as rationals.
+
+    Examples
+    ========
+
+    >>> from sympy import S, symbols, Rational, Float
+    >>> from sympy.core.numbers import equal_valued
+    >>> equal_valued(1, 2)
+    False
+    >>> equal_valued(1, 1)
+    True
+
+    In SymPy expressions with Floats compare unequal to corresponding
+    expressions with rationals:
+
+    >>> x = symbols('x')
+    >>> x**2 == x**2.0
+    False
+
+    However an individual Float compares equal to a Rational:
+
+    >>> Rational(1, 2) == Float(0.5)
+    True
+
+    In a future version of SymPy this might change so that Rational and Float
+    compare unequal. This function is provided in to provide the behaviour
+    currently expected of ``==`` so that it could still be used if the
+    behaviour of ``==`` were to change.
+
+    >>> equal_valued(1, 1.0) # Float vs Rational
+    True
+    >>> equal_valued(S(1).n(3), S(1).n(5)) # Floats of different precision
+    True
+
+    Explanation
+    ===========
+
+    In future SymPy verions Float and Rational might compare unequal and floats
+    with different precisions might compare unequal. In that context a function
+    is needed that can check if a number is equal to 1 or 0 etc. The idea is
+    that instead of testing ``if x == 1:`` if we want to accept floats like
+    ``1.0`` as well then the test can be written as ``if equal_valued(x, 1):``
+    or ``if equal_valued(x, 2):``. Since this function is intended to be used
+    in situations where one or both operands are expected to be concrete
+    numbers like 1 or 0 the function does not recurse through the args of any
+    compound expression to compare any nested floats.
+
+    References
+    ==========
+
+    .. [1] https://github.com/sympy/sympy/pull/20033
+    """
+    x = _sympify(x)
+    y = _sympify(y)
+
+    # Handle everything except Float/Rational first
+    if not x.is_Float and not y.is_Float:
+        return x == y
+    elif x.is_Float and y.is_Float:
+        # Compare values without regard for precision
+        return x._mpf_ == y._mpf_
+    elif x.is_Float:
+        x, y = y, x
+    if not x.is_Rational:
+        return False
+
+    # Now y is Float and x is Rational. A simple approach at this point would
+    # just be x == Rational(y) but if y has a large exponent creating a
+    # Rational could be prohibitively expensive.
+
+    sign, man, exp, _ = y._mpf_
```

Comment:
Can we `return x._as_mpf_val(y._prec) == y._mpf_`?

Reply:
That would not be an exact conversion so it would give slightly different behaviour. For example:
```python
In [126]: x = Rational(1, 3)

In [127]: y = Float(x)

In [128]: x._as_mpf_val(y._prec) == y._mpf_
Out[128]: True
```

Reply:
So is the intent to allow any exact floats (made by division of an integer by a power of 2) equal to the corresponding Rational? It just seems like the numerical wrangling here could be simpler.
```python
p, q=x.p, x.q
t = trailing(q)
return q>>t==1 and y*(1<<t)==p  # or y*(1<<t)-p==0
```
But maybe that 2nd part of the test defeats the purpose of the routine?

Reply:
`y` here is a float so what does `y*(1<<t) == p` do?

Reply:
The logic that I have here in this function is fiddly but it's actually very similar to the existing logic for comparing Float and Rational (just without using `integer_log`):
https://github.com/sympy/sympy/blob/c4d6b3eb1c7b39832116f7b02baf524a1ffadfba/sympy/core/numbers.py#L1891-L1916

Reply:
> so what does `y*(1<<t) == p` do?

It checks whether something like 3.0*(1<<2)==12